kratos/linear_solvers/scaling_solver.h

```diff
@@ -298,226 +301,164 @@ class ScalingSolver
-
-            GS2weights(    number_of_rows,
-                           row_iter_begin,
-                           index_2_begin,
-                           value_begin,
-                           partition[thread_id],
-                           aux
-                      );
-        }
+        IndexPartition<std::size_t>(A.size1()).for_each([&A, &aux](std::size_t Index){
+            ConstIndexIteratorType row_begin_it = A.index1_data().begin() + Index;
+            ConstIndexIteratorType index2_begin_it = A.index2_data().begin() + (*row_begin_it); // Not strictly needed for GS2weights_row, but kept for consistency
+            ConstValueIteratorType value_begin_it = A.value_data().begin() + (*row_begin_it);
+            int number_of_entries_in_row = *(row_begin_it+1) - *row_begin_it;
+
+            GS2weights_row(
+                Index, // Current row index
+                number_of_entries_in_row,
+                row_begin_it,
+                index2_begin_it, // Not used in current GS2weights_row implementation
+                value_begin_it,
+                aux
+            );
+        });
     }
 
-    /**
-     * calculates partial product resetting to Zero the output before
-     */
-    static void GS2weights(
-        int number_of_rows,
-        typename boost::numeric::ublas::compressed_matrix<typename TSparseSpaceType::DataType>::index_array_type::const_iterator row_begin,
-        typename boost::numeric::ublas::compressed_matrix<typename TSparseSpaceType::DataType>::index_array_type::const_iterator index2_begin,
-        typename boost::numeric::ublas::compressed_matrix<typename TSparseSpaceType::DataType>::value_array_type::const_iterator value_begin,
-        unsigned int output_begin_index,
+    static void GS2weights_row(
+        std::size_t kkk, // Current row index
+        int number_of_entries_in_row,
+        ConstIndexIteratorType row_it,
+        ConstIndexIteratorType index2_begin, // Not used in original GS2weights, but kept for consistency if needed later
+        ConstValueIteratorType value_begin,
         VectorType& weights
     )
     {
-        int row_size;
-        typename SparseMatrixType::index_array_type::const_iterator row_it = row_begin;
-        int kkk = output_begin_index;
-        for(int k = 0; k < number_of_rows; k++)
+        double t = 0.0;
+
+        for(int i = 0; i < number_of_entries_in_row; i++)
         {
-            row_size= *(row_it+1)-*row_it;
-            row_it++;
-            double t = 0.0;
-
-            for(int i = 0; i<row_size; i++)
-            {
-                double tmp = std::abs(*value_begin);
-                t += tmp*tmp;
-                value_begin++;
-            }
-            t = sqrt(t);
-            weights[kkk++] = t;
+            double tmp = std::abs(*value_begin);
+            t += tmp*tmp;
+            value_begin++;
         }
+        t = std::sqrt(t); // Corrected from sqrt(t) to std::sqrt(t)
+        weights[kkk] = t;
```

Comment:
this whole thing could be smth like

```cpp
weights[kkk] = std::sqrt(std::inner_product(
    value_begin,
    value_begin + number_of_entries_in_row,
    value_begin,
    1.0
));
```

Reply:
I will check (as I said I tried to minimize modifications to avoid blocking)